File: platforms/apple/Sources/OrbitalVueCore/CatalogModels.swift
import Foundation

public let orbitalVueCatalogContractVersion = "1.0"

public enum ChannelKind: String, Codable, CaseIterable, Sendable {
    case live
    case movie
    case series
    case recording
    case replay
    case music

    public var label: String { rawValue.uppercased() }
}

public enum CatalogSourceType: String, Codable, CaseIterable, Sendable {
    case m3uFile = "m3u-file"
    case m3uURL = "m3u-url"
    case xtream
    case plex
    case emby
    case generated
}

public struct CatalogSource: Codable, Equatable, Sendable, Identifiable {
    public let id: String
    public let name: String
    public let type: CatalogSourceType
    public let displayLocation: String
    public let refreshOnLaunch: Bool

    public init(
        id: String,
        name: String,
        type: CatalogSourceType,
        displayLocation: String,
        refreshOnLaunch: Bool
    ) {
        self.id = id
        self.name = name
        self.type = type
        self.displayLocation = displayLocation
        self.refreshOnLaunch = refreshOnLaunch
    }
}

public struct StreamDescriptor: Codable, Equatable, Sendable {
    public let uri: String
    public let requestHeaders: [String: String]

    public init(uri: String, requestHeaders: [String: String] = [:]) {
        self.uri = uri
        self.requestHeaders = requestHeaders
    }
}

public struct GuideMetadata: Codable, Equatable, Sendable {
    public let tvgId: String?
    public let tvgName: String?
    public let logoUri: String?

    public init(tvgId: String? = nil, tvgName: String? = nil, logoUri: String? = nil) {
        self.tvgId = tvgId
        self.tvgName = tvgName
        self.logoUri = logoUri
    }

    public var isEmpty: Bool {
        tvgId == nil && tvgName == nil && logoUri == nil
    }
}

public struct CatchupMetadata: Codable, Equatable, Sendable {
    public let mode: String
    public let source: String
    public let days: Int
    public let correctionMinutes: Int

    public init(mode: String, source: String, days: Int, correctionMinutes: Int) {
        self.mode = mode
        self.source = source
        self.days = days
        self.correctionMinutes = correctionMinutes
    }
}

public struct CatalogMediaMetadata: Codable, Equatable, Sendable {
    public let libraryId: String?
    public let libraryTitle: String?
    public let seriesTitle: String?
    public let seasonNumber: Int?
    public let episodeNumber: Int?
    public let year: Int?
    public let durationMs: Int?
    public let resumePositionMs: Int?
    public let played: Bool?
    public let addedAt: String?
    public let lastPlayedAt: String?

    public init(
        libraryId: String? = nil,
        libraryTitle: String? = nil,
        seriesTitle: String? = nil,
        seasonNumber: Int? = nil,
        episodeNumber: Int? = nil,
        year: Int? = nil,
        durationMs: Int? = nil,
        resumePositionMs: Int? = nil,
        played: Bool? = nil,
        addedAt: String? = nil,
        lastPlayedAt: String? = nil
    ) {
        self.libraryId = libraryId
        self.libraryTitle = libraryTitle
        self.seriesTitle = seriesTitle
        self.seasonNumber = seasonNumber
        self.episodeNumber = episodeNumber
        self.year = year
        self.durationMs = durationMs
        self.resumePositionMs = resumePositionMs
        self.played = played
        self.addedAt = addedAt
        self.lastPlayedAt = lastPlayedAt
    }
}

public struct CatalogChannel: Codable, Equatable, Sendable, Identifiable {
    public let id: String
    public let number: Int
    public let name: String
    public let group: String
    public let kind: ChannelKind
    public let sourceId: String
    public let stream: StreamDescriptor
    public let guide: GuideMetadata?
    public let catchup: CatchupMetadata?
    public let tags: [String]?
    public let media: CatalogMediaMetadata?

    public init(
        id: String,
        number: Int,
        name: String,
        group: String,
        kind: ChannelKind,
        sourceId: String,
        stream: StreamDescriptor,
        guide: GuideMetadata? = nil,
        catchup: CatchupMetadata? = nil,
        tags: [String]? = nil,
        media: CatalogMediaMetadata? = nil
    ) {
        self.id = id
        self.number = number
        self.name = name
        self.group = group
        self.kind = kind
        self.sourceId = sourceId
        self.stream = stream
        self.guide = guide
        self.catchup = catchup
        self.tags = tags
        self.media = media
    }

    public var initials: String {
        let words = name.split(whereSeparator: \.isWhitespace)
        if words.isEmpty { return "TV" }
        if words.count == 1 { return String(words[0].prefix(2)).uppercased() }
        return words.prefix(2).compactMap(\.first).map(String.init).joined().uppercased()
    }

    public var searchableText: String {
        [
            name,
            group,
            guide?.tvgName ?? "",
            media?.libraryTitle ?? "",
            media?.seriesTitle ?? "",
            media?.year.map(String.init) ?? ""
        ].joined(separator: "\n").uppercased()
    }
}

public struct OrbitalVueCatalog: Codable, Equatable, Sendable {
    public let contractVersion: String
    public let catalogId: String
    public let displayName: String
    public let loadedAt: String
    public let sources: [CatalogSource]
    public let guideSources: [String]
    public let channels: [CatalogChannel]

    public init(
        contractVersion: String = orbitalVueCatalogContractVersion,
        catalogId: String,
        displayName: String,
        loadedAt: String,
        sources: [CatalogSource],
        guideSources: [String],
        channels: [CatalogChannel]
    ) {
        self.contractVersion = contractVersion
        self.catalogId = catalogId
        self.displayName = displayName
        self.loadedAt = loadedAt
        self.sources = sources
        self.guideSources = guideSources
        self.channels = channels
    }

    public var groups: [ChannelGroup] {
        var positions: [String: Int] = [:]
        var values: [ChannelGroup] = []
        for channel in channels {
            if let index = positions[channel.group] {
                values[index].count += 1
            } else {
                positions[channel.group] = values.count
                values.append(ChannelGroup(name: channel.group, count: 1))
            }
        }
        return values
    }
}

public struct ChannelGroup: Equatable, Sendable, Identifiable {
    public let name: String
    public var count: Int
    public var id: String { name }

    public init(name: String, count: Int) {
        self.name = name
        self.count = count
    }
}

public struct ParsedPlaylist: Equatable, Sendable {
    public let channels: [CatalogChannel]
    public let guideSources: [String]

    public init(channels: [CatalogChannel], guideSources: [String]) {
        self.channels = channels
        self.guideSources = guideSources
    }
}

public struct LoadedCatalog: Equatable, Sendable {
    public let catalog: OrbitalVueCatalog
    public let notice: String?
    public let usedCachedFallback: Bool

    public init(catalog: OrbitalVueCatalog, notice: String? = nil, usedCachedFallback: Bool = false) {
        self.catalog = catalog
        self.notice = notice
        self.usedCachedFallback = usedCachedFallback
    }
}

File: platforms/android/app/src/main/java/com/orbitalvue/player/data/CatalogModels.kt
package com.orbitalvue.player.data

import java.time.Duration
import java.time.Instant

enum class ChannelKind(val wireName: String, val label: String) {
    Live("live", "LIVE"),
    Movie("movie", "MOVIE"),
    Series("series", "SERIES"),
    Recording("recording", "RECORDING"),
    Replay("replay", "REPLAY"),
    Music("music", "MUSIC")
}

data class CatchupMetadata(
    val mode: String,
    val source: String,
    val days: Int,
    val correctionMinutes: Int
)

data class Channel(
    val id: String,
    val number: Int,
    val name: String,
    val streamUri: String,
    val group: String = "Uncategorized",
    val logoUri: String? = null,
    val tvgId: String? = null,
    val tvgName: String? = null,
    val requestHeaders: Map<String, String> = emptyMap(),
    val startPositionMs: Long? = null,
    val playbackReportSessionId: String? = null,
    val kind: ChannelKind = ChannelKind.Live,
    val sourceId: String,
    val sourceName: String,
    val catchup: CatchupMetadata? = null,
    val isMediaCenterItem: Boolean = false,
    val libraryId: String? = null,
    val libraryTitle: String? = null,
    val seriesTitle: String? = null,
    val seasonNumber: Int? = null,
    val episodeNumber: Int? = null,
    val year: Int? = null,
    val durationMs: Long? = null,
    val resumePositionMs: Long? = null,
    val played: Boolean = false,
    val addedAt: String? = null,
    val lastPlayedAt: String? = null
) {
    val initials: String
        get() {
            val words = name.trim().split(Regex("\\s+")).filter(String::isNotBlank)
            return when {
                words.isEmpty() -> "TV"
                words.size == 1 -> words.first().take(2).uppercase()
                else -> words.take(2).joinToString("") { it.first().uppercase() }
            }
        }

    val searchText: String
        get() = listOf(
            name,
            group,
            tvgName.orEmpty(),
            sourceName,
            libraryTitle.orEmpty(),
            seriesTitle.orEmpty(),
            year?.toString().orEmpty()
        ).joinToString("\n").uppercase()

    val canResume: Boolean
        get() {
            val resume = resumePositionMs ?: 0
            val duration = durationMs ?: 0
            return isMediaCenterItem && resume >= 30_000 &&
                (duration <= 0 || resume < duration - 30_000)
        }

    val watchProgress: Float?
        get() {
            val duration = durationMs ?: return null
            val resume = resumePositionMs ?: return null
            if (!canResume || duration <= 0) return null
            return (resume.toDouble() / duration.toDouble()).coerceIn(0.0, 1.0).toFloat()
        }

    val watchProgressLabel: String?
        get() {
            val progress = watchProgress ?: return null
            val remainingMinutes = (((durationMs ?: 0) - (resumePositionMs ?: 0)) / 60_000)
                .coerceAtLeast(1)
            return "Continue • ${(progress * 100).toInt()}% • ${formatMinutes(remainingMinutes)} left"
        }

    val mediaMetadataLine: String?
        get() {
            if (!isMediaCenterItem) return null
            val values = ArrayList<String>(3)
            if (kind == ChannelKind.Series && !seriesTitle.isNullOrBlank()) values += seriesTitle
            else if (!libraryTitle.isNullOrBlank()) values += libraryTitle
            year?.takeIf { it in 1888..3000 }?.let { values += it.toString() }
            durationMs?.takeIf { it > 0 }?.let { values += formatDuration(it) }
            return values.takeIf { it.isNotEmpty() }?.joinToString(" • ") ?: group
        }

    private fun formatDuration(milliseconds: Long): String {
        val totalMinutes = (milliseconds / 60_000).coerceAtLeast(1)
        val hours = totalMinutes / 60
        val minutes = totalMinutes % 60
        return if (hours > 0) "${hours}h ${minutes}m" else "${minutes}m"
    }

    private fun formatMinutes(minutes: Long): String {
        val hours = minutes / 60
        val remainder = minutes % 60
        return if (hours > 0) "${hours}h ${remainder}m" else "${minutes}m"
    }
}

enum class MediaLibraryBrowseMode(val label: String) {
    All("All"),
    ContinueWatching("Continue"),
    RecentlyAdded("Recent"),
    Live("Live"),
    Movies("Movies"),
    Series("Series");

    val isEditorial: Boolean
        get() = this == ContinueWatching || this == RecentlyAdded

    val sectionTitle: String
        get() = when (this) {
            ContinueWatching -> "Continue Watching"
            RecentlyAdded -> "Recently Added"
            else -> label
        }
}

data class MediaLibraryBrowseSummary(
    val isMediaCenterLibrary: Boolean = false,
    val continueWatchingCount: Int = 0,
    val recentlyAddedCount: Int = 0,
    val movieCount: Int = 0,
    val seriesCount: Int = 0
)

object MediaLibraryBrowsePolicy {
    private val recentWindow: Duration = Duration.ofDays(30)
    private val futureAllowance: Duration = Duration.ofDays(1)

    fun summarize(
        channels: List<Channel>,
        now: Instant = Instant.now()
    ): MediaLibraryBrowseSummary = MediaLibraryBrowseSummary(
        isMediaCenterLibrary = channels.any(Channel::isMediaCenterItem),
        continueWatchingCount = channels.count { matches(it, MediaLibraryBrowseMode.ContinueWatching, now) },
        recentlyAddedCount = channels.count { matches(it, MediaLibraryBrowseMode.RecentlyAdded, now) },
        movieCount = channels.count { it.kind == ChannelKind.Movie },
        seriesCount = channels.count { it.kind == ChannelKind.Series }
    )

    fun matches(
        channel: Channel,
        mode: MediaLibraryBrowseMode,
        now: Instant = Instant.now()
    ): Boolean = when (mode) {
        MediaLibraryBrowseMode.All -> true
        MediaLibraryBrowseMode.ContinueWatching -> channel.canResume
        MediaLibraryBrowseMode.RecentlyAdded -> channel.isMediaCenterItem &&
            parseInstant(channel.addedAt)?.let { added ->
                !added.isAfter(now.plus(futureAllowance)) && !added.isBefore(now.minus(recentWindow))
            } == true
        MediaLibraryBrowseMode.Live -> channel.kind == ChannelKind.Live
        MediaLibraryBrowseMode.Movies -> channel.kind == ChannelKind.Movie
        MediaLibraryBrowseMode.Series -> channel.kind == ChannelKind.Series
    }

    fun order(channels: List<Channel>, mode: MediaLibraryBrowseMode): List<Channel> = when (mode) {
        MediaLibraryBrowseMode.ContinueWatching -> channels.sortedWith(
            compareByDescending<Channel> { parseInstant(it.lastPlayedAt) ?: Instant.MIN }
                .thenByDescending { it.resumePositionMs ?: 0 }
                .thenBy(Channel::number)
        )
        MediaLibraryBrowseMode.RecentlyAdded -> channels.sortedWith(
            compareByDescending<Channel> { parseInstant(it.addedAt) ?: Instant.MIN }
                .thenBy(Channel::number)
        )
        else -> channels
    }

    private fun parseInstant(value: String?): Instant? = value?.let {
        runCatching { Instant.parse(it) }.getOrNull()
    }
}

data class ParsedPlaylist(
    val channels: List<Channel>,
    val guideSources: List<String>
)

enum class SourceType(val storedValue: String) {
    File("file"),
    Url("url"),
    Plex("plex"),
    Emby("emby");

    val isMediaCenter: Boolean
        get() = this == Plex || this == Emby

    companion object {
        fun fromStored(value: String?): SourceType? = entries.firstOrNull { it.storedValue == value }
    }
}

data class SourceDescriptor(
    val id: String,
    val name: String,
    val type: SourceType,
    val displayLocation: String,
    val refreshOnLaunch: Boolean,
    val usedCachedFallback: Boolean = false
)

data class Catalog(
    val id: String,
    val displayName: String,
    val loadedAt: String,
    val source: SourceDescriptor,
    val guideSources: List<String>,
    val channels: List<Channel>
)

data class LoadedCatalog(
    val catalog: Catalog,
    val notice: String? = null
)
